### PR TITLE
update the data model according recent discussion

### PR DESCRIPTION
**Pull request by YuChaode**
1. add a new root;
2. remove part-number from NE;
3. add component-specific structure;
4. change the configuration capabilities of attributes;
5. add text proposal from Sergio about container;
6. move the section of scalability issue to appendix;

### CODE REVIEW

**Review by sergiobelotti — APPROVED** [review]

File: ietf-network-inventory.tree [hunk 1]
```diff
@@ -1,37 +1,49 @@
...
+            +--ro ne-type?         identityref
+            +--ro uuid?            yang:uuid
+            +--rw name?            string
+            +--rw description?     string
+            +--rw alias?           string
+            +--ro hardware-rev?    string
+            +--ro software-rev?    string
+            +--ro mfg-name?        string
+            +--ro mfg-date?        yang:date-and-time
+            +--ro part-number?     string
```
Comment (sergiobelotti) [inline ietf-network-inventory.tree:15]: Why part-number is still present in the sub-tree of NE?

**Review by sergiobelotti — APPROVED** [review]

File: draft-ietf-ivy-network-inventory-yang.md [hunk 2]
```diff
@@ -265,12 +268,10 @@ in this document uses a flat list of network element.
...
 The "ne-type" is defined as a YANG identity to describe the type of the network element. This document defines only the "physical-network-element" identity.
 
-The component definition is also generalized to support any type of
+The component definition is also generalized to support any types of
 component, such as hardware, software, or firmware.
 
-The component "class" is defined as a union between the hardware
-class identity, defined in "iana-hardware", and the "non-hardware"
-identity, defined in this document.
+Different types of components can be distinguished by the component-sepecific-info structure.
```
[the remark this answers is not in the document]
Comment (YuChaode) [reply draft-ietf-ivy-network-inventory-yang.md]: Thank you Sergio! I changed this sentence to be "Different types of components can be distinguished by the class of component. The component "class" is defined as a union between the hardware class identity, defined in "iana-hardware", and the "non-hardware" identity, defined in this document. Attributes related to specific class of component can be found in the component-specific-info structure." In the previous discussion, Italo suggested to remove class attribute and different types of components can be identified by the component-specific structure. I raised a concern that it is hard to do a filtering if there is not a class attribute. So I prefer to keeping class attribute still in this draft at first, and whetherto keep it or not still needs more discussions.

File: ietf-network-inventory.tree (hunk 1 above)
Comment (sergiobelotti) [reply ietf-network-inventory.tree]: Hi Chaode, yes I know about assembly-id , my suggestion was only on the fact that in YANG modeule part-numer does not appear anymore, but is still there in the tree.  Thanks

File: draft-ietf-ivy-network-inventory-yang.md [hunk 3]
```diff
@@ -475,11 +476,21 @@ The hierarchical components' identifier could be found by the "component-referen
...
+      +--:(chassis)
+      |  +--rw chassis-specific-info
+      +--:(container)
+      |  +--rw slot-specific-info
+      +--:(module)
+      |  +--rw board-specific-info
+      +--:(port)
+      |  +--rw port-specific-info
+      +--:(software)
+         +--rw software-specific-info
```
Comment (italobusi) [inline draft-ietf-ivy-network-inventory-yang.md:493]: The `component-class` is at the same level as `component-id`:

```suggestion
+--rw components
   +--rw component* [component-id]
      +--rw component-id            string
      .......................................
      +--rw (component-class)?
         +--:(chassis)
         |  +--rw chassis-specific-info
         +--:(container)
         |  +--rw slot-specific-info
         +--:(module)
         |  +--rw board-specific-info
         +--:(port)
         |  +--rw port-specific-info
         +--:(software)
            +--rw software-specific-info
```

File: draft-ietf-ivy-network-inventory-yang.md [hunk 4]
```diff
@@ -626,6 +627,18 @@ Mostly, our inventory data model can cover the attributes from OpenConfig.
...
+
+The model proposed by this draft is designed to be as generic as possible so to cover future special types of inventory objects that could be used in other technologies, that have not been identified yet. If the inventory objects were to be defined directly with fixed hierarchical relationships in YANG model, this new type of inventory objects needs to be manually defined, which is not a backward compatible change and therefore is not an acceptable approach for implementation. With a generic model, it is only needed to augment a new component class and extend some specific attributes for this new inventory component class, which is more flexible. We consider that this generic data model, enabling a flexible and backward compatible approach for other technologies, represents the main scope of this draft. Solution description to efficiency/scalability limitations mentioned above is considered as out-of-scope.
+
+
```
Comment (italobusi) [inline draft-ietf-ivy-network-inventory-yang.md:641]: In order to have an appendix number, please move this text before the `{: numbered="false"}` code

File: ietf-network-inventory.yang [hunk 5]
```diff
@@ -2,12 +2,6 @@ module ietf-network-inventory {
   yang-version 1.1;
   namespace "urn:ietf:params:xml:ns:yang:ietf-network-inventory";
   prefix ni;
-
-  import iana-hardware {
-    prefix ianahw;
-    reference
-      "https://www.iana.org/assignments/yang-parameters";
-  }
```
Comment (italobusi) [inline ietf-network-inventory.yang:10]: I think you need to import iana-hardware: it is used by the class data node below

File: ietf-network-inventory.yang [hunk 6]
```diff
@@ -136,79 +131,165 @@ module ietf-network-inventory {
...
-      "The top-level container for the list of network elements 
-      within the network.";
-    list network-element {
-      key ne-id;
+  grouping component-specific-info-grouping {
+    description 
+      "Provide component-specific attributes for different 
+      components.";
+    choice component-class {
+      config false;
```
Comment (italobusi) [inline ietf-network-inventory.yang:174]: Unless we agree that this model can only be used as RO, it is possible that some component-specific attribute might be RW. It is suggested not to preclude this option:

```suggestion
```

**Review by italobusi — CHANGES_REQUESTED** [review]
Few comments to be addressed before merging the PR:
1. Please update the YANG revision date every time the YANG module is updated;
2. See also some detailed review comments within the files changed

Additional comments that may be addressed after the PR is merged (if not addressed, we can track them as new open issues):
1. There is no need for empty grouping for the component-specific attributes. An empty container defined as an augmentation target should be enough for the base network inventory model
2. It is preferable to avoid defining grouping when used once unless this would help the readability of the code

**Review by philbedard — CHANGES_REQUESTED** [review]
Just the clarification on the class section otherwise the rest looked good.

File: ietf-network-inventory.tree (hunk 1 above)
Comment (YuChaode) [reply ietf-network-inventory.tree]: thank you Sergio, now it is removed.

File: draft-ietf-ivy-network-inventory-yang.md (hunk 3 above)
Comment (YuChaode) [reply draft-ietf-ivy-network-inventory-yang.md:493]: Thank you Italo, now it is moved to the same level with component-id.

File: draft-ietf-ivy-network-inventory-yang.md (hunk 4 above)
Comment (YuChaode) [reply draft-ietf-ivy-network-inventory-yang.md]: Fixed.

File: ietf-network-inventory.yang (hunk 5 above)
Comment (YuChaode) [reply ietf-network-inventory.yang]: Thank you, it is added back.

File: ietf-network-inventory.yang (hunk 6 above)
Comment (YuChaode) [reply ietf-network-inventory.yang]: Now this structure is configable. But i have a concern that the change from RW to RO will be a NBC.

File: ietf-network-inventory.yang [hunk 7]
```diff
@@ -136,79 +131,165 @@ module ietf-network-inventory {
...
-  container network-elements {
-    description
-      "The top-level container for the list of network elements 
-      within the network.";
-    list network-element {
-      key ne-id;
+  grouping component-specific-info-grouping {
+    description 
+      "Provide component-specific attributes for different 
+      components.";
```
[the remark this answers is not in the document]
Comment (YuChaode) [reply ietf-network-inventory.yang:178]: Sorry Phil, this structure is actually designed for additional attributes specific to a single type of component.  Just like what we said in the section of 2.4.2.

File: draft-ietf-ivy-network-inventory-yang.md (hunk 4 above)
Comment (italobusi) [reply draft-ietf-ivy-network-inventory-yang.md]: The `{: numbered="false"}` code should be kept for the Acknowledgments section which is not numbered as an appendix

I have also noted that the current md file defines an Appendix with title Appendix which is not needed (the '--- back' indicates that the following paragraphs are appendices)

Proposed changes:

OLD

--- back

```
# Appendix

## Comparison With Openconfig-platform Data Model

<...>

## Efficiency Issue

<...>

# Acknowledgments
```

NEW 

```
--- back

# Comparison With Openconfig-platform Data Model

<...>

# Efficiency Issue

<...>

{: numbered="false"}

# Acknowledgments

```

File: ietf-network-inventory.yang (hunk 6 above)
Comment (italobusi) [reply ietf-network-inventory.yang]: Please note that the NBC change requirement applies only to YANG modules published as IETF RFCs. NBC changes are allowed between different versions of I-Ds.

The issues related to the level of configuration is required for the inventory model are still open: https://github.com/ietf-ivy-wg/network-inventory-yang/issues?q=is%3Aissue+is%3Aopen+label%3Aconfig

File: draft-ietf-ivy-network-inventory-yang.md (hunk 4 above)
Comment (YuChaode) [reply draft-ietf-ivy-network-inventory-yang.md]: Thanks Italo, now it is fixed.

**Review by italobusi — APPROVED** [review]

**Pull request merged**